cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/InMemoryProgramRunDispatcher.java

```diff
@@ -0,0 +1,456 @@
+    RunId runId = programRunInfo.getRunId();
+    LOG.debug("Start pipeline launch for Run ID: {}", runId);
+    ProgramDescriptor programDescriptor = programRunInfo.getProgramDescriptor();
+    ProgramOptions options = programRunInfo.getProgramOptions();
+    isDistributed = programRunInfo.isDistributed();
+    ProgramId programId = programDescriptor.getProgramId();
+    ClusterMode clusterMode = ProgramRunners.getClusterMode(options);
+
+    // Creates the ProgramRunner based on the cluster mode
+    ProgramRunner runner = (clusterMode == ClusterMode.ON_PREMISE ? programRunnerFactory
+        : Optional.ofNullable(remoteProgramRunnerFactory)
+            .orElseThrow(UnsupportedOperationException::new)).create(programId.getType());
+
+    File tempDir = createTempDirectory(programId, runId);
+    AtomicReference<Runnable> cleanUpTaskRef = new AtomicReference<>(
+        createCleanupTask(tempDir, runner));
+
+    // Get the artifact details and save it into the program options.
+    ArtifactId artifactId = programDescriptor.getArtifactId();
+    ArtifactDetail artifactDetail = getArtifactDetail(artifactId);
+    ApplicationSpecification appSpec = programDescriptor.getApplicationSpecification();
+    ProgramDescriptor newProgramDescriptor = programDescriptor;
+
+    boolean isPreview = Boolean.parseBoolean(
+        options.getArguments().getOption(ProgramOptionConstants.IS_PREVIEW, "false"));
+    // do the app spec regeneration if the mode is on premise, for isolated mode, the regeneration
+    // is done on the runtime environment before the program launch
+    // for preview we already have a resolved app spec, so no need to regenerate the app spec again
+    if (!isPreview && appSpec != null && ClusterMode.ON_PREMISE.equals(clusterMode)) {
+      try {
+        ApplicationSpecification generatedAppSpec = regenerateAppSpec(artifactDetail, programId,
+            artifactId, appSpec, options);
+        appSpec = generatedAppSpec != null ? generatedAppSpec : appSpec;
+        newProgramDescriptor = new ProgramDescriptor(programDescriptor.getProgramId(), appSpec);
+      } catch (Exception e) {
+        LOG.warn("Failed to regenerate the app spec for program {}, using the existing app spec",
+            programId);
+      }
+    }
+
+    ProgramOptions runtimeProgramOptions = updateProgramOptions(artifactId, programId, options,
+        runId, clusterMode,
+        Iterables.getFirst(artifactDetail.getMeta().getClasses().getApps(), null));
+
+    // Take a snapshot of all the plugin artifacts used by the program
+    ProgramOptions optionsWithPlugins = createPluginSnapshot(runtimeProgramOptions, programId,
+        tempDir, newProgramDescriptor.getApplicationSpecification());
+
+    // Create and run the program
+    Program executableProgram = createProgram(cConf, runner, newProgramDescriptor, artifactDetail,
+        tempDir);
+    cleanUpTaskRef.set(createCleanupTask(cleanUpTaskRef.get(), executableProgram));
```

Comment:
seems like this is never run anywhere?

Reply:
@albertshau @masoud-io Yes we're not running this cleanup task as of right now. How do we go about cleaning this up now? Since the runtimeInfo is created in app-fabric and previously we were cleaning up after runtimeInfo was closed. Now the program, tempDir and runner are in system pods. So when and how do we clean it up?

Reply:
One approach that I can see is we assign an ID to this cleanup task and when runtimeInfo is closed at app-fabric, we send a request to system pod to execute the cleanupTask corresponding to that ID.

Reply:
We can provide another method in the Dispatcher interface for cleanup. For InMemory implementation, we'll do it directly from app-fabric and for Remote, we can send a request for the same.

Reply:
As discussed with Terence, have added a cleanup task after program run.

Reply:
Seems like this doesn't need to be an AtomicReference anymore, and doesn't need to be created twice. Can just create it once for tempDir, runner, executableProgram and submit the runnable directly.

Reply:
Changed.